citest/json_contract/observation_predicate.py

```diff
@@ -0,0 +1,224 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import re
+
+from ..json_predicate import predicate
+from ..json_predicate import sequenced_predicate_result
+
+
+"""Support for applying predicates to Observations.
+
+These predicates mediate between observations and generic ValuePredicates.
+Specializations target particular attributes of an observation and
+act as a guard to verify that the attribute is set before delegating,
+and calls the delegate with the attribute value rather than the Observation.
+"""
+
+class ExceptionMatchesPredicate(predicate.ValuePredicate):
+  """A predicate that checks a Python exception."""
+
+  @property
+  def klass(self):
+    """The expected exception class."""
+    return self.__klass
+
+  @property
+  def regex(self):
+    """An optional regex to expect in the exception message."""
+    return self.__regex
+
+  def __init__(self, klass, regex=None):
+    self.__klass = klass
+    self.__regex = regex
+
+  def __eq__(self, pred):
+    return (self.__class__ == pred.__class__
+            and self.__regex == pred.regex)
+
+  def __call__(self, context, value):
+    """Implements ValuePredicate interface."""
+    value_class_name = value.__class__.__name__
+    if not isinstance(value, self.__klass):
+      return predicate.PredicateResult(
+          False,
+          comment='Expected {0}, got {1}.'.format(
+              self.__klass.__name__, value_class_name))
+
+    if self.__regex:
+      msg = value.message
+      if re.match(self.__regex, msg):
+        return predicate.PredicateResult(True, comment='Error matches.')
+      else:
+        return predicate.PredicateResult(False, comment='Errors differ.')
+    return predicate.PredicateResult(True, comment='Error matches.')
+
+  def export_to_json_snapshot(self, snapshot, entity):
+    """Implements JsonSnapshotableEntity interface."""
+    snapshot.edge_builder.make_control(
+        entity, 'Class', self.__klass.__name__)
+    if self.__regex:
+      snapshot.edge_builder.make_control(
+          entity, 'Regex', self.__regex)
+
+
+class ObservationPredicate(predicate.ValuePredicate):
+  """A placeholder indicating a type of ObservationPredicate.
+
+  ObservationPredicates are ValuePredicates whose value is
+  expected to be a json_contract.Observation.
+  """
+  pass
+
+
+class ObservationErrorPredicate(ObservationPredicate):
+  """An observation predicate that mediates Observation.errors.
+
+  Note that this mediates individual errors,
+  not the whole list of errors.
```

Comment:
Why is this class constrained to not be able to operate on the whole list, like `ObservationValuePredicate`?